tree/dataframe/src/RSqliteDS.cxx

```diff
@@ -13,19 +13,352 @@
+
+////////////////////////////////////////////////////////////////////////////
+/// Releases the resources associated to a file opened with davix
+int VfsRdOnlyClose(sqlite3_file *pFile)
+{
+   Davix::DavixError *err = nullptr;
+   VfsRootFile *p = reinterpret_cast<VfsRootFile *>(pFile);
+   auto retval = p->pos.close(p->fd, &err);
+   // We can't use delete because the storage for p is managed by sqlite
+   p->~VfsRootFile();
+   if (retval == -1) {
+      return SQLITE_IOERR_CLOSE;
+   }
+   return SQLITE_OK;
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// Issues an HTTP range request to read a chunk from a remote file
+int VfsRdOnlyRead(sqlite3_file *pFile, void *zBuf, int count, sqlite_int64 offset)
+{
+   Davix::DavixError *err = nullptr;
+   VfsRootFile *p = reinterpret_cast<VfsRootFile *>(pFile);
+   if (p->pos.pread(p->fd, zBuf, count, offset, &err) == -1) {
+      return SQLITE_IOERR;
+   }
+   return SQLITE_OK;
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// We do not write to a database in the RDataSource and therefore can simply return an error for this callback
+int VfsRdOnlyWrite(sqlite3_file * /*pFile*/, const void * /*zBuf*/, int /*iAmt*/, sqlite_int64 /*iOfst*/)
+{
+   return SQLITE_OPEN_READONLY;
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// We do not write to a database in the RDataSource and therefore can simply return an error for this callback
+int VfsRdOnlyTruncate(sqlite3_file * /*pFile*/, sqlite_int64 /*size*/)
+{
+   return SQLITE_OPEN_READONLY;
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// As the database is read-only, syncing data to disc is a no-op and always succeeds
+int VfsRdOnlySync(sqlite3_file * /*pFile*/, int /*flags*/)
+{
+   return SQLITE_OK;
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// Returns the cached file size
+int VfsRdOnlyFileSize(sqlite3_file *pFile, sqlite_int64 *pSize)
+{
+   VfsRootFile *p = reinterpret_cast<VfsRootFile *>(pFile);
+   *pSize = p->size;
+   return SQLITE_OK;
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// As the database is read-only, locks for concurrent access are no-ops and always succeeds
+int VfsRdOnlyLock(sqlite3_file * /*pFile*/, int /*level*/)
+{
+   return SQLITE_OK;
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// As the database is read-only, locks for concurrent access are no-ops and always succeeds
+int VfsRdOnlyUnlock(sqlite3_file * /*pFile*/, int /*level*/)
+{
+   return SQLITE_OK;
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// As the database is read-only, locks for concurrent access are no-ops and always succeeds
+int VfsRdOnlyCheckReservedLock(sqlite3_file * /*pFile*/, int *pResOut)
+{
+   *pResOut = 0;
```

Comment:
`nullptr`?

Reply:
No, he's setting the integer (pointed to by `pResOut`) to the value `0`. Same for the others.